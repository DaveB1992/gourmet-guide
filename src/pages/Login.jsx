import React, { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { signIn } from "../config/Firebase";
import { setLoggedIn } from "../config/Auth";
import "../styles/login.css";

const Login = ({ setIsLoggedIn }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [alertMessage, setAlertMessage] = useState("");
  const [isSuccess, setIsSuccess] = useState(false);
  const navigate = useNavigate();

  const handleLogin = async (e) => {
    e.preventDefault();
    const user = await signIn(email, password);
    if (user) {
      setAlertMessage("Login successful!");
      setIsSuccess(true);

      setEmail("");
      setPassword("");

      setLoggedIn();
      setIsLoggedIn(true);

      setTimeout(() => {
        navigate("/cookbook");
      }, 1500);
    } else {
      setAlertMessage("Incorrect email or password.");
      setIsSuccess(false);
    }
  };

  return (
    <div className="login-container">
      <h2>Login</h2>
      {alertMessage && (
        <div className={`alert ${isSuccess ? "success" : "error"}`}>
          {alertMessage}
        </div>
      )}
      <form onSubmit={handleLogin}>
        <div className="form-group">
          <input
            type="email"
            placeholder="Email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
          />
        </div>
        <div className="form-group">
          <input
            type="password"
            placeholder="Password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </div>
        <button type="submit">Login</button>
      </form>
      <p>
        Don&apos;t have an account? <Link to="/sign-up">Sign Up</Link>
      </p>
    </div>
  );
};

export default Login;
