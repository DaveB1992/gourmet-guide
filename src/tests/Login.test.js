import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Login from "../pages/Login";
import * as firebaseMock from "firebase-mock";

describe("Login", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<Login />, { wrapper: MemoryRouter });
    expect(asFragment()).toMatchSnapshot();
  });

  it("allows inputting username and password", () => {
    const { getByPlaceholderText } = render(
      <MemoryRouter>
        <Login setIsLoggedIn={() => {}} />
      </MemoryRouter>
    );

    const emailInput = getByPlaceholderText("Email");
    const passwordInput = getByPlaceholderText("Password");

    fireEvent.change(emailInput, { target: { value: "test@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "test123" } });

    expect(emailInput.value).toBe("test@example.com");
    expect(passwordInput.value).toBe("test123");
  });

  it("displays success message after successful login", async () => {
    const mockAuth = new firebaseMock.MockAuthentication();
    const mockAuthClient = new firebaseMock.MockFirebaseSdk(
      null,
      () => mockAuth,
      null,
      null,
      null
    );

    jest.mock("firebase/app", () => ({
      initializeApp: jest.fn(() => mockAuthClient),
      auth: jest.fn(() => mockAuth),
    }));

    const { getByPlaceholderText, getByText, getByRole } = render(
      <MemoryRouter>
        <Login setIsLoggedIn={() => {}} />
      </MemoryRouter>
    );

    const emailInput = getByPlaceholderText("Email");
    const passwordInput = getByPlaceholderText("Password");
    const submitButton = getByRole("button", { name: "Login" });

    mockAuth.changeAuthState({
      uid: "testUid",
      email: "test1@test.com",
    });

    fireEvent.change(emailInput, { target: { value: "test1@test.com" } });
    fireEvent.change(passwordInput, { target: { value: "password" } });
    fireEvent.click(submitButton);

    await waitFor(() => {
      const successMessage = getByText("Login successful!");
      expect(successMessage).toBeInTheDocument();
    });
  });

  it("displays 'Incorrect email or password.' message after failed login", async () => {
    const mockAuth = new firebaseMock.MockAuthentication();
    const mockAuthClient = new firebaseMock.MockFirebaseSdk(
      null,
      () => mockAuth,
      null,
      null,
      null
    );

    jest.mock("firebase/app", () => ({
      initializeApp: jest.fn(() => mockAuthClient),
      auth: jest.fn(() => mockAuth),
    }));

    const { getByPlaceholderText, getByText, getByRole } = render(
      <MemoryRouter>
        <Login setIsLoggedIn={() => {}} />
      </MemoryRouter>
    );

    const emailInput = getByPlaceholderText("Email");
    const passwordInput = getByPlaceholderText("Password");
    const submitButton = getByRole("button", { name: "Login" });

    mockAuth.changeAuthState(null);

    fireEvent.change(emailInput, { target: { value: "invalid@test.com" } });
    fireEvent.change(passwordInput, { target: { value: "invalidpassword" } });
    fireEvent.click(submitButton);

    await waitFor(() => {
      const errorMessage = getByText("Incorrect email or password.");
      expect(errorMessage).toBeInTheDocument();
    });
  });
});
