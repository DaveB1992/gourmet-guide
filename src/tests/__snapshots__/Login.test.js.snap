// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
<DocumentFragment>
  <div
    class="login-container"
  >
    <div
      class="login-title"
    >
      Login
    </div>
    <form>
      <div
        class="form-group"
      >
        <input
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Login
      </button>
    </form>
    <p
      class="signup-msg"
    >
      Don't have an account? 
      <a
        href="/sign-up"
      >
        Sign Up
      </a>
    </p>
  </div>
</DocumentFragment>
`;
